src/validate-file.test.ts

```diff
@@ -0,0 +1,199 @@
+import { describe, expect, test } from "vitest"
+import { validateFile } from "./validate-file"
+import path from "path"
+
+function fromCwd(relativePath: string) {
+  return path.join(process.cwd(), relativePath)
+}
+
+describe("validateFile", () => {
+  test("denyFiles.client / env:client / failed string validation", () => {
+    expect(() =>
+      validateFile({
+        absolutePath: fromCwd("/lib/server-only.ts"),
+        denyFiles: {
+          client: ["lib/server-only.ts"],
+          server: [],
+        },
+        root: fromCwd("/"),
+        importer: fromCwd("/path/to/importer.ts"),
+        env: "client",
+      })
+    ).toThrowErrorMatchingInlineSnapshot(
+      `
+    [Error: File denied in client environment
+     - File: lib/server-only.ts
+     - Importer: path/to/importer.ts
+     - Matcher: "lib/server-only.ts"]
```

Comment:
Can we rename `Matcher` to `Pattern` since this isn't an _agent_ that takes action, but rather a piece of data to declaratively tell the matcher what to do

Reply:
Done.